[clang-tidy] Improved --verify-config when using literal style in config file

felix642 opened the pull request:
Specifying checks using the literal style (|) in the clang-tidy config file is currently supported but was not implemented for the --verify-config options. This means that clang-tidy would work properly but, using the --verify-config option would raise an error due to some checks not being parsed properly.

Fixes #53737

CC @SimplyDanny

felix642 committed:
[clang-tidy] Improved --verify-config when using literal style in config file

Specifying checks using the literal style (|) in the clang-tidy config file is currently
supported but was not implemented for the --verify-config options.
This means that clang-tidy would work properly but, using the --verify-config option
would raise an error due to some checks not being parsed properly.

Fixes #53737

SimplyDanny reviewed (approved):
Seems like the parsing logic got duplicated. That eventually caused the misbehavior in verification. Alignment would be preferred.

From my point of view, the fix is okay for now though.

felix642 committed:
fixup! [clang-tidy] Improved --verify-config when using literal style in config file

Partially rewrote verifyChecks() to realign the checks parsing logic.

Instead of manually parsing the string of globs, verifyCHecks() now uses a GlobList which is responsible
to parse the string and return a list of regexes. This ensures that any changes made to the checks parsing logic
has to be made at only one place rather than two.

felix642 committed:
fixup! [clang-tidy] Improved --verify-config when using literal style in config file

Fixed small type in release notes

felix642 commented:
Thank you for the review @SimplyDanny, as per your suggestion, I've rewrote part of my fix to unify the parsing logic rather than updating the duplicated code. The method now uses the GlobList which already handles correctly the regexes generation from the list of checks.

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -262,6 +262,9 @@ Miscellaneous
   option is specified. Now ``clang-apply-replacements`` applies formatting only with
   the option.
 
+- Fixed ``--verify-check`` option not properly parsing checks when using the 
+  literal operator in the ``.clang-tidy`` config.
```
I think this should be in line 103.

SimplyDanny commented on clang-tools-extra/clang-tidy/GlobList.cpp:
```diff
@@ -19,12 +19,17 @@ static bool consumeNegativeIndicator(StringRef &GlobList) {
-// Converts first glob from the comma-separated list of globs to Regex and
-// removes it and the trailing comma from the GlobList.
-static llvm::Regex consumeGlob(StringRef &GlobList) {
+// Extract the first glob from the comma-separated list of globs
```
```suggestion
// Extracts the first glob from the comma-separated list of globs,
```

SimplyDanny commented on clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp:
```diff
@@ -454,52 +454,31 @@ static constexpr StringLiteral VerifyConfigWarningEnd = " [-verify-config]\n";
-                        [&Glob](StringRef S) { return Glob.match(S); })) {
-        AnyInvalid = true;
+    if (llvm::none_of(AllChecks.keys(), [&Reg](StringRef S) {
+          llvm::errs() << S << '\n';
```
Is this output intended?

felix642 replied:
No, I forgot to remove this print before committing, it should be fixed now. Thank you

SimplyDanny commented on clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp:
```diff
@@ -454,52 +454,31 @@ static constexpr StringLiteral VerifyConfigWarningEnd = " [-verify-config]\n";
-    if (Cur.empty())
+  for (const auto &Item : Globs.getItems()) {
+    const llvm::Regex &Reg = Item.Regex;
+    const llvm::StringRef Text = Item.Text;
```
Fine from my point of view to inline both of these variables as it doesn't improve much to have them. If you insist to keep them, use them wherever `Item.` appears.

SimplyDanny commented on clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp:
```diff
@@ -454,52 +454,31 @@ static constexpr StringLiteral VerifyConfigWarningEnd = " [-verify-config]\n";
+          return Reg.match(S);
+        })) {
+      AnyInvalid = true;
+      if (Item.Text.contains('*'))
```
Previously this came before the checks for matches in line 466. Seems like the semantic has so changed slightly if only performance-wise, hasn't it?

felix642 replied:
You are right, I've changed slightly this function's semantics, but it shouldn't have any observable behaviour change for the end user.
Before my change, we would only create a regex and try to match it if the string contains any '*' character. But, since the GlobList implicitly generates a list of Regexes we always try to match the strings through regexp now. This won't make a difference since the generated regex will only match the exact string if it doesn't contain any '*', but it will be a bit less efficient since we are going through pattern matching rather than string comparison.

I've moved the condition with the '*' character at the of the method to preserve the error message that is displayed to the user in case we can't find any matching check. (Either with "check glob ... doesn't match" if it contains a '*' or "unknown check ..." if it doesn't).

SimplyDanny replied:
Fine for me. Yet, as I'm not that proficient with the detailed goals of the project, I'd like to ask @njames93 or @AaronBallman if this potentially slight performance decrease is acceptable.

PiotrZSL replied:
This code is called only on --verify-config, performance is not a issue here.

felix642 committed:
fixup! [clang-tidy] Improved --verify-config when using literal style in config file

Fixed minor issues found in code review

felix642 committed:
fixup! [clang-tidy] Improved --verify-config when using literal style in config file

Code formatting

PiotrZSL reviewed (approved):
LGTM
